functions/zoom-downloader.py

```diff
@@ -565,41 +575,52 @@ def stream_file_to_s3(self):
             {"uploading file to S3": self.s3_filename,
              "metadata": metadata}
         )
-        part_info = {"Parts": []}
-        mpu = s3.create_multipart_upload(
+        parts = []
+        mpu = self.s3.create_multipart_upload(
                     Bucket=ZOOM_VIDEOS_BUCKET,
                     Key=self.s3_filename,
                     Metadata=metadata)
+        self.upload_id = mpu["UploadId"]
 
         try:
             chunks = enumerate(
                         self.stream.iter_content(chunk_size=MIN_CHUNK_SIZE), 1
                      )
-            for part_number, chunk in chunks:
-                part = s3.upload_part(Body=chunk,
-                                      Bucket=ZOOM_VIDEOS_BUCKET,
-                                      Key=self.s3_filename,
-                                      PartNumber=part_number,
-                                      UploadId=mpu["UploadId"])
-
-                part_info["Parts"].append({
-                    "PartNumber": part_number,
-                    "ETag": part["ETag"]
-                })
-
-            s3.complete_multipart_upload(Bucket=ZOOM_VIDEOS_BUCKET,
-                                         Key=self.s3_filename,
-                                         UploadId=mpu["UploadId"],
-                                         MultipartUpload=part_info)
+            with concurrent.futures.ThreadPoolExecutor() as executor:
+                future_map = {}
+                for part_number, chunk in chunks:
+                    f = executor.submit(self.upload_part, part_number, chunk)
+                    future_map[f] = part_number
+                
+                for future in concurrent.futures.as_completed(future_map):
+                    part_number = future_map[future]
+                    part = future.result()
+                    parts.append({
+                        "PartNumber": part_number,
+                        "ETag": part["ETag"]
+                    })
+
+            # complete_multipart_upload requires parts in order by part number
+            parts = sorted(parts, key=lambda i: i["PartNumber"])
+
+            self.s3.complete_multipart_upload(
+                Bucket=ZOOM_VIDEOS_BUCKET,
+                Key=self.s3_filename,
+                UploadId=mpu["UploadId"],
+                MultipartUpload={"Parts": parts}
+            )
             print("Completed multipart upload of {}.".format(self.s3_filename))
         except Exception as e:
             logger.exception(
                 "Something went wrong with upload of {}:{}"
                 .format(self.s3_filename, e)
             )
-            s3.abort_multipart_upload(Bucket=ZOOM_VIDEOS_BUCKET,
-                                      Key=self.s3_filename,
-                                      UploadId=mpu["UploadId"])
+            self.s3.abort_multipart_upload(
```

Comment:
I wonder a little about if there are ways for the concurrent.futures stuff to go wrong that wouldn't end up here. Just thinking out loud, not asking for changes. I guess something could get hung and the whole lambda times out (which already can happen). Wasn't there something we had to do once related to half-finished multi-part bucket files?

Reply:
We have a deletion policy for parts uploaded and abandoned by a failed multipart upload. I think that's what you are thinking of?